x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_installation/perform_rule_installation_handler.ts

```diff
@@ -0,0 +1,145 @@
+  context: SecuritySolutionRequestHandlerContext,
+  request: KibanaRequest<unknown, unknown, PerformRuleInstallationRequestBody>,
+  response: KibanaResponseFactory
+) => {
+  const siemResponse = buildSiemResponse(response);
+
+  try {
+    const ctx = await context.resolve(['core', 'alerting', 'securitySolution']);
+    const config = ctx.securitySolution.getConfig();
+    const soClient = ctx.core.savedObjects.client;
+    const rulesClient = await ctx.alerting.getRulesClient();
+    const detectionRulesClient = ctx.securitySolution.getDetectionRulesClient();
+    const ruleAssetsClient = createPrebuiltRuleAssetsClient(soClient);
+    const ruleObjectsClient = createPrebuiltRuleObjectsClient(rulesClient);
+    const exceptionsListClient = ctx.securitySolution.getExceptionListClient();
+
+    const { mode } = request.body;
+
+    // This will create the endpoint list if it does not exist yet
+    await exceptionsListClient?.createEndpointList();
+
+    // If this API is used directly without hitting any detection engine
+    // pages first, the rules package might be missing.
+    await ensureLatestRulesPackageInstalled(ruleAssetsClient, config, ctx.securitySolution);
+
+    const allLatestVersions = await ruleAssetsClient.fetchLatestVersions();
+    const currentRuleVersions = await ruleObjectsClient.fetchInstalledRuleVersions();
+    const currentRuleVersionsMap = new Map(
+      currentRuleVersions.map((version) => [version.rule_id, version])
+    );
+
+    const allInstallableRules = allLatestVersions.filter((latestVersion) => {
+      const currentVersion = currentRuleVersionsMap.get(latestVersion.rule_id);
+      return !currentVersion;
+    });
+
+    const ruleInstallQueue: Array<{
+      rule_id: RuleSignatureId;
+      version: RuleVersion;
+    }> = [];
+    const ruleErrors = [];
+    const installedRules = [];
+    const skippedRules: SkippedRuleInstall[] = [];
+
+    // Perform all the checks we can before we start the upgrade process
+    if (mode === 'SPECIFIC_RULES') {
+      const installableRuleIds = new Set(allInstallableRules.map((rule) => rule.rule_id));
+      request.body.rules.forEach((rule) => {
+        // Check that the requested rule is not installed yet
+        if (currentRuleVersionsMap.has(rule.rule_id)) {
+          skippedRules.push({
+            rule_id: rule.rule_id,
+            reason: SkipRuleInstallReason.ALREADY_INSTALLED,
+          });
+          return;
+        }
+
+        // Check that the requested rule is installable
+        if (!installableRuleIds.has(rule.rule_id)) {
+          ruleErrors.push({
+            error: new Error(
+              `Rule with ID "${rule.rule_id}" and version "${rule.version}" not found`
+            ),
+            item: rule,
+          });
+          return;
+        }
+
+        ruleInstallQueue.push(rule);
+      });
+    } else if (mode === 'ALL_RULES') {
+      ruleInstallQueue.push(...allInstallableRules);
+    }
```

Comment:
nit: It looks like this code chunk could be moved to a separate function returning `ruleInstallQueue`.